Introduce TLS client-side connection

n-ramacciotti opened the pull request:
This PR introduces initial TLS client functionality and provides tool for testing client connections

### Dependency

This PR depends on the following PRs: vsock support and mmio native support . Those commits are included here for testing.

### Implementation Details

The implementation is based on [rustls crate](https://docs.rs/rustls/latest/rustls/), which in no_std environment requires to implement `CryptoProvider` and `TimeProvider`. Due to the absence of a trusted time source,`TimeProvider` is implemented as a fixed source of time. The `CryptoProvider` is based on the example provided in the rustls repository for the moment.

Rustls offers an unbuffered API for `no_std`, which requires the user to manage all the buffers for TLS and plaintext data. The struct `TlsBuffer` was created for this purpose.

Other struct were introduced to help manage the TLS connection. In particular, `TlsClient` provides a way to build a `TlsConnection`,  which handles handshake, sending and receiving of application records and closing the connection (The core of the functionalities).

The `CryptoProvider` needs a random number generator. Currently, the getrandom crate with feature `custom` is used, but this is just a temporary solution. 

### Compilation
To compile the feature `tls` needs to be enabled.

A CA certificate must be present at compilation time.
If the cert is missing, the `insert_ca_cert()` function in `kernel/build.rs` calls the script `scripts/gen_ca_server_certs.sh`, which generates signed certificates for both a CA and a server. It is possible to run `scripts/gen_ca_server_certs.sh` directly to run tests.

### Testing
Two example function are provided and are commented inside `kernel/src/svsm.rs`. Uncomment one to use it:
- `test_https()`: does a single GET request to a web server(i.e. `openssl s_server -WWW ...`) and then closes the connection. To simplify the testing process, a simple script automates certificates creation, compilation and launch of socat  and openssl on the host.
- `test_command_server()`: connects to a python server (`tls_command_server/command_server.py`) from which it can receive "commands" to execute. For the moment are simple messages. It stops when it receives the `EXIT` command.

#### Notes
- The scripts provides a certificate starting from the current date. The time provider needs to provide a date within the certificate validity.
- The stack size needed to be increased due to a double panic during the call of some rustls api. 
- To work with the examples, on the host run:
  - `socat -d -d   VSOCK-LISTEN:12345,reuseaddr,fork   TCP:localhost:4433`
  - start a server (openssl or the python server) 
 - `TlsConnection`, with read and write methods, provides support for a single TLS record at the time. So if an application record is divided in more TLS records more read or write operations are required.

--- 
For review: @stefano-garzarella @osteffenrh @luigix25

luigix25 committed:
Revert "virtio: remove vsock support"

Commit c4719d3 dropped driver support for virtio-vsock because it was not being used.

This reverts this commit because it will be used for attestion.

This revert does not add back the examples/aarch64/main.rs file as it is not needed.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/vsock: introduce virtio-vsock support

Add a virtio-vsock wrapper around virtio-drivers. It provides blocking functions
for connect, send, recv and close.

There is also a VsockStream implemetation that offers Read and Write traits.

To use it you need both the `vsock` and the `virtio-drivers` features.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
scripts/launch_guest: add vsock to launch_guest

Add an optional --vsock parameter. This will attach a vsock-device
to the VM.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
virtio-drivers/connectiomanager: store connection status in `Connection`

Status of the connection is not stored in the `Connection` struct. This can cause
weird behaviours like a `send` not failing after a `shutdown`.

Introduce the `status` field that keeps track of the current status of the connection.
This field is checked before any operation.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
virtio-drivers/connectiomanager: introduce get_connection_status

Introduce `get_connection_status` function that returns the current status of
the connection.

This can be used, for example, to implement some blocking wrappers for functions
like `connect`.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
aproxy: Use read/write traits

Change parameters types from UnixStream to Read/Write traits.
This allows to call these functions using a different type
that implements those traits.

No behaviour changes intended.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
aproxy: enable vsock for attestation

Two new parameters: `vsock_cid` and `vsock_port` to specify the CID of the SVSM's device
and the port that it is using.

Note that Serial and Vsock are mutually exclusive.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel: add in-svsm tests for VirtioVsockDriver

Add a test that performs some basic checks on vsock functionalities:

- double connection
- double close
- send a buffer after connection is close
- recv a buffer from the host

The vsock server is created on the host using nc.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/attest: abstract transport implementation

Introduce a `Transport` enum that implements the read and write operations
according to the specific type.

This is to create an abstraction over the transport method to support other
transports, like virtio-vsock, without too many code changes.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/attest: use vsock when feature is available

Use vsock for attestation when the `vsock` feature is enabled.

If vsock fails it will fallback on the serial port.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/fw_cfg: Add get_vsock_attest_port

Add `get_vsock_attest_port`, this function can be used to get the
value of the fw_cfg option `opt/org.svsm/VsockAttestPort` that changes
the default vsock port used for attestation.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

luigix25 committed:
kernel/vsock: send credit update

virtio-vsock spec forbids to send data if the receiver's queue is full.
If the guest is expecting some more data from the host and the host is waiting
for some space in the buffer, this leads to a deadlock.

Send a credit update to inform the peer about the space available in the queue.

Signed-off-by: Luigi Leonardi <leonardi@redhat.com>

n-ramacciotti committed:
kernel/tls: add client examples and testing tools

- Implement basic TLS clients (https and simple command interaction)
- Provide script to launch a get request to an openssl s_server (HTTPS)
- Add a python based server to test interactive command client (persistent connection)

Note: Currently, both clients need socat and the server running on the host.

n-ramacciotti committed:
kernel: introduce TLS client support

- Add `CryptoProvider` and `TimeProvider`: required by no_std rustls
to ensure cryptographic security and access to the current time
- Implement `TlsBuffer`: stores and manages records necessary for
the TLS.
- Include `KeyLogger`: useful for debugging TLS keys
- Include error handling
- Provide custom RNG: required to generate secure secret (temporary
solution, based on getrandom crate)
- Add `tls` feature: enable the functionality only when needed
- Add certificate generation script: simple way to make certificates
(with current date as certificate start date)

Note:
- `TimeProvider` provides a fixed timestamp (real-time not yet
 available), check the validity of the dates in the certificate
before testing

luigix25 reviewed (commented):
I added a few comments, I suggest you to enable the CI here so that it runs some checks for you.

for example `make clippy` fails when it does not find the root CA or your commits are missing the `Signed-off-by`

n-ramacciotti commented on kernel/src/svsm.rs:
```diff
@@ -66,6 +66,9 @@ use kbs_types::Tee;
 #[cfg(feature = "tls")]
 use getrandom::register_custom_getrandom;
 #[cfg(feature = "tls")]
+#[allow(unused_imports)]
```
I used it to avoid having to comment the import as well every time I switch function, but since the function is commented I could comment both

n-ramacciotti commented on kernel/Cargo.toml:
```diff
@@ -67,14 +67,27 @@ verify_proof = { workspace = true, optional = true}
-default = []
+tls = ["dep:rustls", "dep:getrandom", "vsock", "virtio-drivers", "rustcrypto"]
+rustcrypto = ["dep:hmac", "dep:ecdsa", "dep:der", "dep:p256", "dep:pkcs8", "dep:rsa", "dep:signature", "dep:x25519-dalek", "dep:rand_core","dep:aes"]
+default = ["tls"]
```
If you mean `tls` as `default` feature, then yes, I forgot to remove it. If you're referring to the division of `tls` and `rustcrypto`, I did it to reduce column width and also to highlight that `tls` is based on some crates from RustCrypto

n-ramacciotti commented on kernel/src/mm/address_space.rs:
```diff
@@ -159,7 +159,7 @@ pub const SIZE_LEVEL1: usize = 1usize << ((9 * 1) + 12);
 
 // Stack definitions
-pub const STACK_PAGES: usize = 8;
+pub const STACK_PAGES: usize = 16;
```
That sounds perfectly reasonable

n-ramacciotti committed:
kernel/mm: increase stack size

Double the stack size: this prevent double panic when creating a TLS connection and is a temporary solution to bypass the problem.
